### PR TITLE
RC_Channel/AP_Arming: add flight mode conflict check

### PR DESCRIPTION
**Pull request by tatsuy**
If the user set the following incorrect settings, he/she cannot change the flight mode from the transmitter.

 FLTMODE_CH 5
 RC5_OPTION 4

I added arming check and error message because assigning any option to the flight mode channel causes unexpected behavior.
I received reports of this issue from two users.

This is the test result in SITL.

```
STABILIZE> param set RC5_OPTION 41
STABILIZE> arm throttle
STABILIZE> Got COMMAND_ACK: COMPONENT_ARM_DISARM: FAILED
APM: PreArm: FLTMODE_CH and RC5_OPTION are in conflict

STABILIZE> param set RC5_OPTION 0
STABILIZE> arm throttle
STABILIZE> Got COMMAND_ACK: COMPONENT_ARM_DISARM: ACCEPTED
APM: Arming motors
ARMED
STABILIZE> APM: Disarming motors
DISARMED

STABILIZE> param set FLTMODE_CH 7
STABILIZE> param show RC7_OPTION
STABILIZE> RC7_OPTION       7.000000

STABILIZE> arm throttle
STABILIZE> Got COMMAND_ACK: COMPONENT_ARM_DISARM: FAILED
APM: PreArm: FLTMODE_CH and RC7_OPTION are in conflict
```

### CODE REVIEW

**Comment by rmackay9** [comment]
Thanks for this.  I wonder why changing modes can't be done if a function has been assigned to the flight mode channel.  It seems like a bad idea to set the flight mode channels' RCx_OPTION to a non-zero value but I wonder why it doesn't work.

@peterbarker I guess this is really your area right?

File: libraries/AP_Arming/AP_Arming.cpp [hunk 1]
```diff
@@ -578,6 +578,11 @@ bool AP_Arming::rc_arm_checks(AP_Arming::Method method)
         check_failed(ARMING_CHECK_PARAMETERS, true, "Duplicate Aux Switch Options");
         check_passed = false;
     }
+    uint8_t flight_mode_channel;
+    if (rc().check_flight_mode_channel_conflict(flight_mode_channel)) {
+        check_failed(ARMING_CHECK_PARAMETERS, true, "FLTMODE_CH and RC%d_OPTION are in conflict", flight_mode_channel);
```
Comment (rmackay9) [inline libraries/AP_Arming/AP_Arming.cpp:583]: I think we could shorten the message to, "FLTMODE_CH and RCx_OPTION conflict".  Also unfortunately Rover's parameter is just MODE_CH so we need to check the build here sadly.

Comment (tatsuy) [reply libraries/AP_Arming/AP_Arming.cpp:583]: How about "Mode channel and RC%d_OPTION conflict"?

**Comment by peterbarker** [comment]
> Thanks for this. I wonder why changing modes can't be done if a function has been assigned to the flight mode channel. It seems like a bad idea to set the flight mode channels' RCx_OPTION to a non-zero value but I wonder why it doesn't work.
> 
> @peterbarker I guess this is really your area right?

Terrible idea to share the channel,and just an artifact of having multiple parameters that can specify the same channel input.  Perhaps for 4.2 we can eliminate Copter's `FLTMODE_CH` and Rover's `MODE_CH` and have `AUX_FUNC::MODE_CHANNEL` instead.

If I were to hazard a guess as to why it doesn't work - probably to do with the debouncing stuff which someone unified between RC Channel Options and the flight-mode-channel about 18 months back IIRC.

File: libraries/AP_Arming/AP_Arming.cpp (hunk 1 above)
Comment (rmackay9) [reply libraries/AP_Arming/AP_Arming.cpp:583]: @tatsuy, yes that sounds good to me. txs!

**Review by rmackay9 — APPROVED** [review]
I've reviewed this and it looks good to me but @peterbarker should decide.

File: libraries/AP_Arming/AP_Arming.cpp [hunk 2]
```diff
@@ -578,6 +578,11 @@ bool AP_Arming::rc_arm_checks(AP_Arming::Method method)
         check_failed(ARMING_CHECK_PARAMETERS, true, "Duplicate Aux Switch Options");
         check_passed = false;
     }
+    uint8_t flight_mode_channel;
+    if (rc().check_flight_mode_channel_conflict(flight_mode_channel)) {
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp:582]:
```suggestion
    if (rc().flight_mode_channel_conflicts_with_rc_option()) {
```

And just use `flight_mode_channel_number()` here.

File: libraries/RC_Channel/RC_Channels.cpp [hunk 3]
```diff
@@ -217,6 +217,22 @@ void RC_Channels::read_mode_switch()
...
+// return true if assigned
+bool RC_Channels::check_flight_mode_channel_conflict(uint8_t &num)
+{
+    num = flight_mode_channel_number();
+    const RC_Channel *c = channel(num-1);
+    if (c == nullptr) {
+        return false;
+    }
+    const uint16_t option = c->option.get();
+    if (option != 0) {
```
Comment (peterbarker) [inline libraries/RC_Channel/RC_Channels.cpp:230]: This should be using the enumeration value, even if a little casting is required (cast `option`, not the enumeration value)

Comment (peterbarker) [reply libraries/RC_Channel/RC_Channels.cpp:230]: I'd also invert this to be `==` to preserve the `return false;`-early pattern.

File: libraries/RC_Channel/RC_Channel.h [hunk 4]
```diff
@@ -418,6 +418,7 @@ class RC_Channels {
     // mode switch handling
     void reset_mode_switch();
     virtual void read_mode_switch();
+    bool check_flight_mode_channel_conflict(uint8_t &mode);
```
Comment (peterbarker) [inline libraries/RC_Channel/RC_Channel.h:421]:
```suggestion
    bool check_flight_mode_channel_conflict(uint8_t &mode);
```

Should be `const`

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_Arming/AP_Arming.cpp (hunk 2 above)
Comment (tatsuy) [reply libraries/AP_Arming/AP_Arming.cpp]: the function is private.
https://github.com/ArduPilot/ardupilot/blob/master/libraries/RC_Channel/RC_Channel.h#L546
Can I change it to public?

File: libraries/RC_Channel/RC_Channels.cpp (hunk 3 above)
Comment (tatsuy) [reply libraries/RC_Channel/RC_Channels.cpp]: thank you.
I fixed it.

File: libraries/AP_Arming/AP_Arming.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_Arming/AP_Arming.cpp]: Yes, that's all good.  We're hoping to remove the `FLTMODE_CH` and `MODE_CH` parameters from the vehicles in 4.2 - but a method to retrieve that channel number will still be good.

Comment (tatsuy) [reply libraries/AP_Arming/AP_Arming.cpp]: I fixed it

File: libraries/RC_Channel/RC_Channel.h (hunk 4 above)
Comment (tatsuy) [reply libraries/RC_Channel/RC_Channel.h]: I created another commit to add const.
please review it.

**Comment by tatsuy** [comment]
@peterbarker 
Thank you for the review.
I confirmed that it works in SITL after my changes.

**Review by peterbarker — APPROVED** [review]
LGTM

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**